boofuzz/blocks/repeat.py

```diff
@@ -91,57 +84,47 @@ def fuzzable(self):
     def original_value(self):
         return self._original_value
 
-    def mutate(self):
-        """
-        Mutate the primitive by stepping through the fuzz library, return False on completion. If variable-bounding is
-        specified then fuzzing is implicitly disabled. Instead, the render() routine will properly calculate the
-        correct repetition and return the appropriate data.
-
-        @rtype:  bool
-        @return: True on success, False otherwise.
-        """
-
-        # render the contents of the block we are repeating.
-        self.request.names[self.block_name].render()
-
+    def mutations(self):
         # if the target block for this sizer is not closed, raise an exception.
         if self.block_name not in self.request.closed_blocks:
             raise exception.SullyRuntimeError("Can't apply repeater to unclosed block: %s" % self.block_name)
 
-        # if we've run out of mutations, raise the completion flag.
-        if self.mutant_index == self.num_mutations():
-            self._fuzz_complete = True
-
-        # if fuzzing was disabled or complete, and mutate() is called, ensure the original value is restored.
-        if not self.fuzzable or self._fuzz_complete:
-            self._value = self.original_value
-            self.current_reps = self.min_reps
-            return False
-
-        if self.variable:
-            self.current_reps = self.variable.render()
+        if not self.fuzzable:
+            return
+        elif self.variable is not None:
+            return  # no need to mutate if the variable block is driving mutations
         else:
-            self.current_reps = self._fuzz_library[self.mutant_index]
-
-        # set the current value as a multiple of the rendered block based on the current fuzz library count.
-        block = self.request.closed_blocks[self.block_name]
-        self._value = block.render() * self._fuzz_library[self.mutant_index]
+            for fuzzed_reps_number in self._fuzz_library:
+                yield Mutation(mutations={self.qualified_name: fuzzed_reps_number})
 
-        # increment the mutation count.
-        self._mutant_index += 1
-
-        return True
-
-    def num_mutations(self):
+    def num_mutations(self, default_value):
         """
         Determine the number of repetitions we will be making.
 
         @rtype:  int
         @return: Number of mutated forms this primitive can take.
+        :param default_value:
         """
 
         return len(self._fuzz_library)
 
+    def encode(self, value, mutation_context):
+        return value * child_data
```

Comment:
Good job for the mutation refactor. Currently I don't know much about the minds behind it. But it should be cool.

Just according to the code, the `child_data` is missing, which is placed in another function. Something maybe like this:

```
def encode(self, value, mutation_context):
    child_data = self._get_child_data(mutation_context=mutation_context)
    return value * child
```

Also, the 1st arg of `_get_child_data()` in this file should be mutation_context instead of mutation.

Reply:
Yep, that's about what it should be. Now that I look at it, Repeat has a few changes that should be made. Repeat wasn't used in my ad hoc testing, so it probably is non-functional right now. I'll take this as a reminder to revisit the whole class.

Reply:
E.g. `Repeat.render_mutated()` can be deleted. `Repeat.render()` will probably get folded into `Repeat._get_child_data()`.

Reply:
Yeah, the `Repeat.render_mutated()` calls `self.encode()` to return a encoded value, which is what `self.encode()` does.